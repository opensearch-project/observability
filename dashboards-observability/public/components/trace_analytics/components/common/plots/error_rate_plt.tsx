/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { EuiHorizontalRule, EuiPanel } from '@elastic/eui';
import moment from 'moment';
import React, { useMemo } from 'react';
import { Plt } from '../../../../visualizations/plotly/plot';
import {
  fixedIntervalToMilli,
  fixedIntervalToTickFormat,
  NoMatchMessage,
  PanelTitle,
} from '../helper_functions';

export function ErrorRatePlt(props: {
  items: { items: Plotly.Data[]; fixedInterval: string };
  setStartTime: (startTime: string) => void;
  setEndTime: (endTime: string) => void;
}) {
  const getLayout = () =>
    ({
      height: 217,
      margin: {
        l: 57,
        r: 5,
        b: 50,
        t: 30,
        pad: 4,
      },
      annotations: props.items.items.length > 0 && [
        {
          x: props.items.items[0]?.x[props.items.items[0]?.x.length - 1],
          y: 0,
          showarrow: true,
          arrowhead: 0,
          xref: 'x',
          yref: 'y',
          text: `Now: ${props.items.items[0]?.y[props.items.items[0]?.y.length - 1]}%`,
          ax: 0,
          ay: -160,
          borderpad: 10,
          arrowwidth: 0.7,
        },
      ],
      showlegend: false,
      xaxis: {
        fixedrange: true,
        showgrid: false,
        visible: true,
        type: 'date',
        tickformat: fixedIntervalToTickFormat(props.items.fixedInterval),
        color: '#899195',
      },
      yaxis: {
        title: {
          text: 'Error rate (%)',
          font: {
            size: 12,
          },
          standoff: 10,
        },
        range: [
          0,
          Math.min(100, Math.max(...(props.items.items[0]?.y.map((y) => y * 1.2) || []), 1)),
        ],
        fixedrange: true,
        ticksuffix: '%',
        gridcolor: '#d9d9d9',
        showgrid: true,
        visible: true,
        color: '#899195',
      },
    } as Partial<Plotly.Layout>);

  const layout = useMemo(() => getLayout(), [props.items]);

  const onClick = (event: any) => {
    if (!event?.points) return;
    const point = event.points[0];
    const start = point.data.x[point.pointNumber];
    const end = start + fixedIntervalToMilli(props.items.fixedInterval);
    props.setStartTime(moment(start).toISOString());
    props.setEndTime(moment(end).toISOString());
    window.scrollTo({ left: 0, top: 0, behavior: 'smooth' });
  };

  return (
    <>
      <EuiPanel style={{ minWidth: 433, minHeight: 308 }}>
        <PanelTitle title="Trace error rate over time" />
        <EuiHorizontalRule margin="m" />
        {props.items?.items?.length > 0 ? (
          <Plt data={props.items.items} layout={layout} onClickHandler={onClick} />
        ) : (
          <NoMatchMessage size="s" />
        )}
      </EuiPanel>
    </>
  );
}
