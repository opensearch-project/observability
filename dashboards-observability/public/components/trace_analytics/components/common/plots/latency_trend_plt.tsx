/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

/*
 *   Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import _ from 'lodash';
import React, { useMemo } from 'react';
import { Plt } from '../../../../visualizations/plotly/plot';

export function LinePlt(props: { data: Plotly.Data[] }) {
  const maxY = props.data[0]?.y ? Math.max(...props.data[0].y) : 0;
  const layout = useMemo(
    () => ({
      plot_bgcolor: 'rgba(0, 0, 0, 0)',
      paper_bgcolor: 'rgba(0, 0, 0, 0)',
      xaxis: {
        fixedrange: true,
        showgrid: false,
        visible: false,
      },
      yaxis: {
        fixedrange: true,
        showgrid: false,
        visible: false,
        range: [0, maxY * 1.1],
      },
      margin: {
        l: 0,
        r: 0,
        b: 0,
        t: 0,
        pad: 0,
      },
      height: 20,
      width: 60,
    }),
    [props.data]
  );
  return props.data[0].x.length > 1 ? <Plt data={props.data} layout={layout} /> : <div>-</div>;
}

export function LatencyPlt(props: { data: Plotly.Data[] }) {
  const layout = useMemo(
    () =>
      ({
        xaxis: {
          showgrid: false,
          type: 'date',
          tickmode: 'auto',
          color: '#899195',
        },
        yaxis: {
          title: {
            text: 'Hourly latency (ms)',
            font: {
              size: 12,
            },
          },
          gridcolor: '#d9d9d9',
          color: '#899195',
        },
        annotations: [
          {
            x: props.data[0].x[props.data[0].x.length - 1],
            y: 0,
            showarrow: true,
            arrowhead: 0,
            xref: 'x',
            yref: 'y',
            text: `Now: ${_.round(props.data[0].y[props.data[0].y.length - 1] as number, 2)}ms`,
            ax: 0,
            ay: -140,
            borderpad: 10,
            arrowwidth: 0.7,
          },
        ],
        margin: {
          l: 50,
          r: 30,
          b: 50,
          t: 30,
          pad: 0,
        },
        height: 200,
        width: 400,
      } as Partial<Plotly.Layout>),
    [props.data]
  );
  return (
    <>
      <Plt data={props.data} layout={layout} />
    </>
  );
}
