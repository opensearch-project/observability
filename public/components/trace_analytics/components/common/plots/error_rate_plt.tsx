/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

/*
 *   Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import { EuiHorizontalRule, EuiPanel } from '@elastic/eui';
import moment from 'moment';
import React, { useMemo } from 'react';
import { Plt } from '../../../../visualizations/plotly/plot';
import {
  fixedIntervalToMilli,
  fixedIntervalToTickFormat,
  NoMatchMessage,
  PanelTitle,
} from '../helper_functions';

export function ErrorRatePlt(props: {
  items: { items: Plotly.Data[]; fixedInterval: string };
  setStartTime: (startTime: string) => void;
  setEndTime: (endTime: string) => void;
}) {
  const getLayout = () =>
    ({
      height: 217,
      margin: {
        l: 57,
        r: 5,
        b: 50,
        t: 30,
        pad: 4,
      },
      annotations: props.items.items.length > 0 && [
        {
          x: props.items.items[0]?.x[props.items.items[0]?.x.length - 1],
          y: 0,
          showarrow: true,
          arrowhead: 0,
          xref: 'x',
          yref: 'y',
          text: `Now: ${props.items.items[0]?.y[props.items.items[0]?.y.length - 1]}%`,
          ax: 0,
          ay: -160,
          borderpad: 10,
          arrowwidth: 0.7,
        },
      ],
      showlegend: false,
      xaxis: {
        fixedrange: true,
        showgrid: false,
        visible: true,
        type: 'date',
        tickformat: fixedIntervalToTickFormat(props.items.fixedInterval),
        color: '#899195',
      },
      yaxis: {
        title: {
          text: 'Error rate (%)',
          font: {
            size: 12,
          },
          standoff: 10,
        },
        range: [
          0,
          Math.min(100, Math.max(...(props.items.items[0]?.y.map((y) => y * 1.2) || []), 1)),
        ],
        fixedrange: true,
        ticksuffix: '%',
        gridcolor: '#d9d9d9',
        showgrid: true,
        visible: true,
        color: '#899195',
      },
    } as Partial<Plotly.Layout>);

  const layout = useMemo(() => getLayout(), [props.items]);

  const onClick = (event: any) => {
    if (!event?.points) return;
    const point = event.points[0];
    const start = point.data.x[point.pointNumber];
    const end = start + fixedIntervalToMilli(props.items.fixedInterval);
    props.setStartTime(moment(start).toISOString());
    props.setEndTime(moment(end).toISOString());
    window.scrollTo({ left: 0, top: 0, behavior: 'smooth' });
  };

  return (
    <>
      <EuiPanel style={{ minWidth: 433, minHeight: 308 }}>
        <PanelTitle title="Trace error rate over time" />
        <EuiHorizontalRule margin="m" />
        {props.items?.items?.length > 0 ? (
          <Plt data={props.items.items} layout={layout} onClickHandler={onClick} />
        ) : (
          <NoMatchMessage size="s" />
        )}
      </EuiPanel>
    </>
  );
}
