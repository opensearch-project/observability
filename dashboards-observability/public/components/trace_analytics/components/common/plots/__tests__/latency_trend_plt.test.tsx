/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import { render } from '@testing-library/react';
import { configure, mount, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { ErrorRatePlt } from '../error_rate_plt';
import { LatencyPlt, LinePlt } from '../latency_trend_plt';

describe('Latency trend plot component', () => {
  configure({ adapter: new Adapter() });

  it('renders line plot', () => {
    const data = [
      {
        x: [1605027600000, 1605027700000],
        y: [78.16, 100],
        type: 'scatter',
        mode: 'lines',
        hoverinfo: 'none',
        line: {
          color: '#000000',
          width: 1,
        },
      },
    ] as Plotly.Data[];
    const wrapper = shallow(<LinePlt data={data} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders dash', () => {
    const data = [
      {
        x: [1605027600000],
        y: [78.16],
        type: 'scatter',
        mode: 'lines',
        hoverinfo: 'none',
        line: {
          color: '#000000',
          width: 1,
        },
      },
    ] as Plotly.Data[];
    const wrapper = shallow(<LinePlt data={data} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders latency plot', () => {
    const data = [
      {
        x: [1605027600000],
        y: [284.47],
        type: 'scatter',
        mode: 'lines+markers',
        hovertemplate: '%{x}<br>Average latency: %{y}<extra></extra>',
        hoverlabel: {
          bgcolor: '#d7c2ff',
        },
        marker: {
          color: '#987dcb',
          size: 8,
        },
        line: {
          color: '#987dcb',
        },
      },
    ] as Plotly.Data[];
    const wrapper = shallow(<LatencyPlt data={data} />);

    expect(wrapper).toMatchSnapshot();
  });
});
