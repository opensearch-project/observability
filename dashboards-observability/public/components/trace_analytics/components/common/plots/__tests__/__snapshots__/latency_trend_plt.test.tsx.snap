// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Latency trend plot component renders dash 1`] = `
<div>
  -
</div>
`;

exports[`Latency trend plot component renders latency plot 1`] = `
<Fragment>
  <Plt
    data={
      Array [
        Object {
          "hoverlabel": Object {
            "bgcolor": "#d7c2ff",
          },
          "hovertemplate": "%{x}<br>Average latency: %{y}<extra></extra>",
          "line": Object {
            "color": "#987dcb",
          },
          "marker": Object {
            "color": "#987dcb",
            "size": 8,
          },
          "mode": "lines+markers",
          "type": "scatter",
          "x": Array [
            1605027600000,
          ],
          "y": Array [
            284.47,
          ],
        },
      ]
    }
    layout={
      Object {
        "annotations": Array [
          Object {
            "arrowhead": 0,
            "arrowwidth": 0.7,
            "ax": 0,
            "ay": -140,
            "borderpad": 10,
            "showarrow": true,
            "text": "Now: 284.47ms",
            "x": 1605027600000,
            "xref": "x",
            "y": 0,
            "yref": "y",
          },
        ],
        "height": 200,
        "margin": Object {
          "b": 50,
          "l": 50,
          "pad": 0,
          "r": 30,
          "t": 30,
        },
        "width": 400,
        "xaxis": Object {
          "color": "#899195",
          "showgrid": false,
          "tickmode": "auto",
          "type": "date",
        },
        "yaxis": Object {
          "color": "#899195",
          "gridcolor": "#d9d9d9",
          "title": Object {
            "font": Object {
              "size": 12,
            },
            "text": "Hourly latency (ms)",
          },
        },
      }
    }
  />
</Fragment>
`;

exports[`Latency trend plot component renders line plot 1`] = `
<Plt
  data={
    Array [
      Object {
        "hoverinfo": "none",
        "line": Object {
          "color": "#000000",
          "width": 1,
        },
        "mode": "lines",
        "type": "scatter",
        "x": Array [
          1605027600000,
          1605027700000,
        ],
        "y": Array [
          78.16,
          100,
        ],
      },
    ]
  }
  layout={
    Object {
      "height": 20,
      "margin": Object {
        "b": 0,
        "l": 0,
        "pad": 0,
        "r": 0,
        "t": 0,
      },
      "paper_bgcolor": "rgba(0, 0, 0, 0)",
      "plot_bgcolor": "rgba(0, 0, 0, 0)",
      "width": 60,
      "xaxis": Object {
        "fixedrange": true,
        "showgrid": false,
        "visible": false,
      },
      "yaxis": Object {
        "fixedrange": true,
        "range": Array [
          0,
          110.00000000000001,
        ],
        "showgrid": false,
        "visible": false,
      },
    }
  }
/>
`;
